app/scripts/lib/AbstractPetnamesBridge.ts

```diff
@@ -0,0 +1,256 @@
+    messenger: PetnamesBridgeMessenger;
+  }) {
+    this.#isTwoWay = isTwoWay;
+    this.#nameController = nameController;
+    this.#messenger = messenger;
+  }
+
+  // Initializes listeners
+  init(): void {
+    if (this.#isTwoWay) {
+      this.#messenger.subscribe('NameController:stateChange', () =>
+        this.#synchronize(SyncDirection.PETNAMES_TO_SOURCE),
+      );
+    }
+    this.onSourceChange(() =>
+      this.#synchronize(SyncDirection.SOURCE_TO_PETNAMES),
+    );
+  }
+
+  /**
+   * Adds a listener for source change events.
+   *
+   * @param listener - The listener function to be called when a source change event occurs.
+   */
+  protected abstract onSourceChange(listener: () => void): void;
+
+  /**
+   * Retrieves the source entries.
+   *
+   * @returns An array of PetnameEntry objects representing the source entries.
+   */
+  protected abstract getSourceEntries(): PetnameEntry[];
+
+  /**
+   * Update the Source with the given entry. To be overridden by two-way subclasses.
+   *
+   * @param type
+   * @param entry
+   */
+  // eslint-disable-next-line @typescript-eslint/no-unused-vars
+  protected updateSourceEntry(type: ChangeType, entry: PetnameEntry): void {
+    throw new Error('updateSourceEntry must be overridden for two-way bridges');
+  }
+
+  /**
+   * Synchronizes Petnames with the Source or vice versa, depending on the direction.
+   *
+   * @param direction - The direction to synchronize in.
+   */
+  #synchronize(direction: SyncDirection): void {
+    if (this.#synchronizingDirection === direction) {
+      throw new Error(
+        `Attempted to synchronize recursively in same direction: ${direction}`,
+      );
+    }
+    if (this.#synchronizingDirection !== null) {
+      return; // Ignore calls while updating in the opposite direction
+    }
+
+    this.#synchronizingDirection = direction;
+
+    const [newEntries, prevEntries] =
+      direction === 'Source->Petnames'
+        ? [this.getSourceEntries(), this.#getPetnameEntries()]
+        : [this.#getPetnameEntries(), this.getSourceEntries()];
+
+    const changeList = this.#computeChangeList(prevEntries, newEntries);
+    this.#applyChangeList(changeList);
+
+    this.#synchronizingDirection = null;
+  }
+
+  /**
+   * Extract PetnameEntry objects from the name controller state.
+   */
+  #getPetnameEntries(): PetnameEntry[] {
+    const { names } = this.#nameController.state;
+    const entries: PetnameEntry[] = [];
+    for (const type of Object.values(NameType)) {
+      for (const value of Object.keys(names[type])) {
+        for (const variation of Object.keys(names[type][value])) {
+          const entry = names[type][value][variation];
+          if (!entry.name) {
+            continue;
+          }
+          entries.push({
+            value,
+            type,
+            name: entry.name,
+            variation,
+            sourceId: entry.sourceId ?? undefined,
+            origin: entry.origin ?? undefined,
```

Comment:
Just to confirm, we're now including the origin in both the source and target entries, but we're not yet filtering the name controller entries based on the origin to support synchronising the delete of the account identities?

This could be done using a new optional `getFilteredTargetEntries` abstract method that can be overridden to filter the current name controller entires in each concrete bridge before comparison. In the case of the account identity bridge, we could just filter by origin.

But I definitely vote to do that in a subsequent PR so we can get this beautiful but big PR in first 😄

Reply:
Forgive me, but I can't find these changes?

I'm expecting some filtering logic in `AccountIdentitiesPetnamesBrdige` in a new abstract / override method?

Reply:
I apologize for the confusion caused by my previous message. Please disregard it. I've moved the changes to https://github.com/MetaMask/metamask-extension/pull/22600.